src/librados/AioCompletionImpl.h

```diff
@@ -59,94 +58,78 @@ struct librados::AioCompletionImpl {
 			io(NULL), aio_write_seq(0), aio_write_list_item(this) { }
 
   int set_complete_callback(void *cb_arg, rados_callback_t cb) {
-    lock.Lock();
+    std::scoped_lock l{lock};
     callback_complete = cb;
     callback_complete_arg = cb_arg;
-    lock.Unlock();
     return 0;
   }
   int set_safe_callback(void *cb_arg, rados_callback_t cb) {
-    lock.Lock();
+    std::scoped_lock l{lock};
     callback_safe = cb;
     callback_safe_arg = cb_arg;
-    lock.Unlock();
     return 0;
   }
   int wait_for_complete() {
-    lock.Lock();
-    while (!complete)
-      cond.Wait(lock);
-    lock.Unlock();
+    std::unique_lock l{lock};
+    cond.wait(l, [this] { return complete; });
     return 0;
   }
   int wait_for_safe() {
     return wait_for_complete();
   }
   int is_complete() {
-    lock.Lock();
-    int r = complete;
-    lock.Unlock();
-    return r;
+    std::scoped_lock l{lock};
+    return complete;
   }
   int is_safe() {
     return is_complete();
   }
   int wait_for_complete_and_cb() {
-    lock.Lock();
-    while (!complete || callback_complete || callback_safe)
-      cond.Wait(lock);
-    lock.Unlock();
+    std::unique_lock l{lock};
+    cond.wait(l, [this] { return complete && !callback_complete && !callback_safe; });
     return 0;
   }
   int wait_for_safe_and_cb() {
     return wait_for_complete_and_cb();
   }
   int is_complete_and_cb() {
-    lock.Lock();
-    int r = complete && !callback_complete && !callback_safe;
-    lock.Unlock();
-    return r;
+    std::scoped_lock l{lock};
+    return complete && !callback_complete && !callback_safe;
   }
   int is_safe_and_cb() {
     return is_complete_and_cb();
   }
   int get_return_value() {
-    lock.Lock();
-    int r = rval;
-    lock.Unlock();
-    return r;
+    std::scoped_lock l{lock};
+    return rval;
   }
   uint64_t get_version() {
-    lock.Lock();
-    version_t v = objver;
-    lock.Unlock();
-    return v;
+    std::scoped_lock l{lock};
+    return objver;
   }
 
   void get() {
-    lock.Lock();
+    std::scoped_lock l{lock};
     _get();
-    lock.Unlock();
   }
   void _get() {
-    ceph_assert(lock.is_locked());
+    // ceph_assert(ceph_mutex_is_locked(lock));
```

Comment:
are we removing this assert? why?

Reply:
the reason is that these locks are constructed like
```c++
lock("AioCompletionImpl lock", false, false)
```
and Mutex' constructor looks like
```c++
Mutex(const std::string &n, bool r = false, bool ld=true, bool bt=false);
```
while the new `ceph::mutex` 's constructor looks like
```c++
mutex_debug_impl(std::string group, bool bt = false)
```

so, we cannot construct a `ceph::mutex` without lockdep support. but if we want to persist these `assert`s,  we would need to use `ceph::mutex` for its `is_locked()` method.

in this specific case, yes, i am removing this assert. but after a second thought, i think we need to keep this `assert()` as they are helpful for debugging.

Reply:
@neha-ojha i added a commit to enable us to construct `ceph::mutex` without lockdep. so i think it's not a problem anymore.